src/caf/core/data_structures.py

```diff
@@ -0,0 +1,655 @@
+            factors_fn = self._hour_to_week_factors
+        elif self == TimeFormat.AVG_HOUR and to_time_format == TimeFormat.AVG_DAY:
+            factors_fn = self._hour_to_day_factors
+        else:
+            raise TypeError(
+                "Cannot figure out the conversion factors to get from "
+                f"time_format {self.value} to {to_time_format.value}"
+            )
+
+        return factors_fn()
+
+
+class DVector:
+    """
+    Class to store and manipulate data with segmentation and optionally zoning.
+
+    The segmentation is stored as an attribute as well as forming the index of
+    the data. Zoning, if present, is stored as an attribute as well as forming
+    the columns of the data. Data is in the form of a dataframe and reads/writes
+    to h5 along with all metadata.
+    """
+
+    _val_col = "val"
+
+    def __init__(
+        self,
+        segmentation: Segmentation,
+        import_data: Union[pd.DataFrame, dict],
+        zoning_system: Optional[ZoningSystem] = None,
+        time_format: Optional[Union[str, TimeFormat]] = None,
+        val_col: Optional[str] = None,
+    ) -> None:
+        """
+        segmentation: An instance of the segmentation class. This should usually
+        be built from enumerated options in the SegmentsSuper class, but custom
+        segments can be user defined if necesssary.
+        import_data: The DVector data. This should usually be a dataframe or path
+        to a dataframe, but there is also an option to read in and convert
+        DVectors in the old format from NorMITs-demand.
+        zoning_system: Instance of ZoningSystem. This must match import data.
+        If this is given, import data must contain zone info in the column names,
+        if this is not given import data must contain only 1 column.
+        """
+        if zoning_system is not None:
+            if not isinstance(zoning_system, ZoningSystem):
+                raise ValueError(
+                    "Given zoning_system is not a caf.core.ZoningSystem object."
+                    f"Got a {type(zoning_system)} object instead."
+                )
+
+        if not isinstance(segmentation, Segmentation):
+            raise ValueError(
+                "Given segmentation is not a caf.core.SegmentationLevel object."
+                f"Got a {type(segmentation)} object instead."
+            )
+
+        self._zoning_system = zoning_system
+        self._segmentation = segmentation
+        self._time_format = self._validate_time_format(time_format)
+
+        # Set defaults if args not set
+        val_col = self._val_col if val_col is None else val_col
+
+        # Try to convert the given data into DVector format
+        if isinstance(import_data, pd.DataFrame) or isinstance(import_data, pd.Series):
+            self._data = self._dataframe_to_dvec(import_data)
+        elif isinstance(import_data, dict):
+            self._data = self._old_to_new_dvec(import_data)
```

Comment:
Do you mean as a class method but still called here, or as a class method and the init only accepts dataframes?

Reply:
class method and the init only accepts dataframes, I think it will make the init clearer and simpler and I don't think people will need the dict functionality very often anyway. Tbh it's mostly my preference I like having an init with a single defined way of initialising the class then any other conversions from other datatypes can just have there own class method e.g. `from_dict`. This is one of the comments that doesn't necessarily need changing.